File: Robotics Programming Training/KPP/Final.cpp
#include <iostream>
#include <vector>
#include <string>

class Map {
private:
    std::vector<std::vector<char>> grid;
    int rows, cols;

public:
    Map(const std::vector<std::string>& input_grid) 
        : rows(input_grid.size()), cols(input_grid[0].length()) {
        grid.resize(rows, std::vector<char>(cols));
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                grid[i][j] = input_grid[i][j];
            }
        }
    }

    std::pair<int, int> get_start() const {
        for (int i = 0; i < rows; ++i) {
            for (int j = 0; j < cols; ++j) {
                if (grid[i][j] == 'S') {
                    return {i, j};
                }
            }
        }
        return {-1, -1};
    }

    void update_cell(int x, int y, char value) {
        grid[x][y] = value;
    }

    int get_rows() const { return rows; }
    int get_cols() const { return cols; }
    char get_cell(int x, int y) const { return grid[x][y]; }
};

class Robot {
private:
    int x, y;
    int time;
    int speed;
    bool can_break_wall;

public:
    Robot(int start_x, int start_y) 
        : x(start_x), y(start_y), time(0), speed(1), can_break_wall(false) {}

    std::string move(char direction, Map& map_obj) {
        int dx = 0, dy = 0;
        if (direction == 'D') dx = 1;
        else if (direction == 'U') dx = -1;
        else if (direction == 'R') dy = 1;
        else if (direction == 'L') dy = -1;

        int new_x = x + dx, new_y = y + dy;

        if (0 <= new_x && new_x < map_obj.get_rows() && 0 <= new_y && new_y < map_obj.get_cols()) {
            char cell = map_obj.get_cell(new_x, new_y);
            if (cell != 'x' || can_break_wall) {
                std::string result = handle_cell(cell);

                x = new_x;
                y = new_y;

                time += (speed == 1) ? 2 : 1;

                if (cell == 'O') time *= 2;
                else if (cell == 'E') speed = 2;
                else if (cell == 'M') can_break_wall = true;
                map_obj.update_cell(new_x, new_y, '.');

                return result;
            } else {
                return "Robot nabrak, silahkan diperbaiki";
            }
        }
        return "Robot mencoba keluar dari peta";
    }

    std::string handle_cell(char cell) {
        switch (cell) {
            case 'M': return "Bertemu dengan mekanik, siap membasmi rintangan";
            case 'E': return "Bertemu dengan electrical, kecepatan roda naik menjadi 200%";
            case 'P': return "Hi programmer bau bawang :3";
            case 'O': return "Bertemu dengan official, diajak ngonten bareng";
            case 'F': return "Robot berhasil mencapai tujuan";
            default: return "";
        }
    }

    int get_time() const { return time; }
};

void simulate_robot(Map& map_obj, const std::string& movements) {
    auto start = map_obj.get_start();
    Robot robot(start.first, start.second);
    bool reached_goal = false;

    for (char move : movements) {
        if (move == 'D' || move == 'R' || move == 'U' || move == 'L') {
            std::string result = robot.move(move, map_obj);
            if (!result.empty()) {
                std::cout << result << std::endl;
            }
            if (result == "Robot berhasil mencapai tujuan") {
                reached_goal = true;
                break;
            }
        }
    }

    if (!reached_goal) {
        std::cout << "Robot gagal dalam mencapai tujuan :(" << std::endl;
    }

    std::cout << "Robot telah berjalan selama " << robot.get_time() << " menit" << std::endl;
}

int main() {
    int M, N;
    std::cin >> M >> N;
    std::cin.ignore();  // Consume newline

    std::vector<std::string> grid(M);
    for (int i = 0; i < M; ++i) {
        std::getline(std::cin, grid[i]);
    }

    Map map_obj(grid);

    int P;
    std::cin >> P;
    std::cin.ignore();  // Consume newline

    std::string movements;
    std::getline(std::cin, movements);

    simulate_robot(map_obj, movements);

    return 0;
}

File: Robotics Programming Training/KPP/final.py
class Robot:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.time = 0
        self.speed = 1
        self.can_break_wall = False
    
    def move(self, direction, map_obj):
        dx, dy = 0, 0
        if direction == 'D':
            dx = 1
        elif direction == 'U':
            dx = -1
        elif direction == 'R':
            dy = 1
        elif direction == 'L':
            dy = -1
        
        new_x, new_y = self.x + dx, self.y + dy
        
        if 0 <= new_x < map_obj.rows and 0 <= new_y < map_obj.cols:
            cell = map_obj.grid[new_x][new_y]
            if cell != 'x' or self.can_break_wall:
                result = self.handle_cell(cell)
                
                # Update posisi robot
                self.x, self.y = new_x, new_y
                
                # Update waktu berdasarkan kecepatan
                if self.speed == 1:
                    self.time += 2 # 2 menit per pindahan untuk speed = 1
                else:
                    self.time += 1
                
                if cell == 'O':
                    self.time *= 2
                elif cell == 'E':
                    self.speed = 2
                elif cell == 'M':
                    self.can_break_wall = True
                map_obj.update_cell(new_x, new_y, '.')

                return result
            else:
                return "Robot nabrak, silahkan diperbaiki"
        
        return "Robot mencoba keluar dari peta"

    def handle_cell(self, cell):
        if cell == 'M':
            return "Bertemu dengan mekanik, siap membasmi rintangan"
        elif cell == 'E':
            return "Bertemu dengan electrical, kecepatan roda naik menjadi 200%"
        elif cell == 'P':
            return "Hi programmer bau bawang :3"
        elif cell == 'O':
            return "Bertemu dengan official, diajak ngonten bareng"
        elif cell == 'F':
            return "Robot berhasil mencapai tujuan"
        return ""

class Map:
    def __init__(self, grid):
        self.grid = [list(row) for row in grid]
        self.rows = len(grid)
        self.cols = len(grid[0])

    def get_start(self):
        for i in range(self.rows):
            for j in range(self.cols):
                if self.grid[i][j] == 'S':
                    return (i, j)
        return None

    def update_cell(self, x, y, value):
        self.grid[x][y] = value

def simulate_robot(map_obj, movements):
    start = map_obj.get_start()
    robot = Robot(start[0], start[1])
    reached_goal = False
    for move in movements:
       if move in ['D', 'R', 'U', 'L']:
            result = robot.move(move, map_obj)
            if result:
                print(result)
            if result == "Robot berhasil mencapai tujuan":
                reached_goal = True
                break
    if not reached_goal:
        print("Robot gagal dalam mencapai tujuan :(")
    
    print(f"Robot telah berjalan selama {int(robot.time)} menit")

# Main execution
M, N = map(int, input().split())
grid = [input().strip() for _ in range(M)]
map_obj = Map(grid)
P = int(input())
movements = input().strip()
simulate_robot(map_obj, movements)

    # ... (rest of the Robot class remains the same)

# ... (rest of the code remains the same)
